### PR TITLE
feat: add `timeout` function

### PR DESCRIPTION
**Pull request by MarlonPassos-git**
> [!TIP]
> The owner of this PR can publish a _preview release_ by commenting `/publish` in this PR. Afterwards, anyone can try it out by running `pnpm add radashi@pr<PR_NUMBER>`.

## Summary

This PR introduces a timeout promise function similar to the delay function, but with the ability to reject upon timeout. It also supports custom error classes passing a function as a parameter.

```ts
// Rejects after 1 second with the default "timeout" error message
await _.timeout(1000)

// Rejects after 1 second with a custom error message
await _.timeout(1000, 'Custom timeout message')

// Rejects after 1 second with a custom error object
await _.timeout(1000, () => new Error('Custom error'))
```

## Related issue, if any:

#203 

## For any code change,

- [x] Related documentation has been updated, if needed
- [x] Related tests have been added or updated, if needed
- [x] Related benchmarks have been added or updated, if needed

## Does this PR introduce a breaking change?

No

## Bundle impact

| Status | File | Size [^1337] |
| --- | --- | --- |
| A | `src/async/TimeoutError.ts` | 127 |
| A | `src/async/timeout.ts` | 246 |

[^1337]: Function size includes the `import` dependencies of the function.

### CODE REVIEW

File: tests/async/timout.test.ts
Comment (SaidbekAbdiganiev) [inline tests/async/timout.test.ts:1]: The `timeout.test.ts` is misspelled as `timout.test.ts` 🙃

**Comment by aleclarson** [comment]
Hey! There's a new requirement for PRs that introduce new features. Without this requirement met, we won't be able to merge this. Note that this PR can still be included in a `beta` prerelease before this requirement is fulfilled.

- [ ] Edit the [next-minor.md](https://github.com/radashi-org/radashi/blob/main/.github/next-minor.md) file

⚠️ *Note: You need to run `git rebase main` before this file will appear locally.*

**Comment by MarlonPassos-git** [comment]
@aleclarson about the `next-minor.md` file, you can check if this is the expected pattern. Maybe we can add some examples in `contributing.md` to make it easier for new contributors

**Comment by aleclarson** [comment]
> Maybe we can add some examples in `contributing.md` to make it easier for new contributors

Great idea 👍 

This was the expected format:

```md
#### timeout

https://github.com/radashi-org/radashi/pull/250
```

But using `#250` is also acceptable.

The idea is that a script will parse that out and fetch the documentation for each function.

If a PR is extending an existing function, I don't know if that trick will work, so the PR author might be on the hook for describing the new feature in `next-minor.md` etc.

**Comment by aleclarson** [comment]
We don't benchmark async functions (or at least, we haven't been).

**Comment by aleclarson** [comment]
Thanks for everything, Marlon! 🚀

**Pull request merged**